File: twosum/2sum.ml
(*two sum problem*)
type Res = (int*int list);
type Index = (int*int) list;

(*the following is a helper function to get the item from a list*)
let rec ElementRetrieve (numList:int list)(lengthOfList:int):int = 
  match numList with
  | [] -> -1;
  |hd::tl -> if lengthOfList=0 then hd else ElementRetrieve tl lengthOfList-1;
;;
(*helper function to retrieve*)
let TwoSum (nums:int list)(target:int):int list =
  (*creating a mutable variable using `ref`*)
  let compliment = ref 0 in
  (*creating a mutable dictionary using ref*)
  let indices:Index = ref {} in
  let numLen = List.length nums in (*length of the list*)
  
  for i = 0 to numLen () do
    compliment = target - ElementRetrieve (nums)(i)

  done
;;

File: twosum/2sum.py
from typing import List

def twoSum(nums: List[int], target: int) -> List[int]:
    indices = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in indices:
            return [indices[complement], i]
        else:
            indices[num] = i
    return []

print(twoSum([1, 2, 3], 5))
